### PR TITLE
Adding vhost vdpa support

### PR DESCRIPTION
**Pull request by qingwangrh**
1. Improve VDPABlkSimulatorTest to support virtio-vdpa and vhost-vdpa
2. Point to  new VirtioVdpaSimulatorTest in vdpa_sim_blk_test

ID:1391

### CODE REVIEW

**Comment by qingwangrh** [comment]
python ConfigTest.py --testcase=vdpa_sim_blk_test --guestname=RHEL.9.3.0 --firmware=default_bios --clone=no --driveformat=virtio_scsi
 (1/3) Host_RHEL.m9.u3.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.3.0.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.multi_disks.q35: STARTED
 (1/3) Host_RHEL.m9.u3.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.3.0.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.multi_disks.q35: PASS (42.73 s)
 (2/3) Host_RHEL.m9.u3.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.3.0.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.hotplug_unplug.q35: STARTED
 (2/3) Host_RHEL.m9.u3.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.3.0.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.hotplug_unplug.q35: PASS (54.16 s)
 (3/3) Host_RHEL.m9.u3.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.3.0.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.discard_check.q35: STARTED
 (3/3) Host_RHEL.m9.u3.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.3.0.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.discard_check.q35: PASS (8.25 s)
RESULTS    : PASS 3 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

**Comment by qingwangrh** [comment]
@XueqiangWei @peixiu Please help to review,thanks.

**Comment by peixiu** [comment]
Hi @qingwangrh , hit some failed on win11 and win2022 guests:

python ConfigTest.py --testcase=vdpa_sim_blk_test --guestname=Win11,Win2022 --clone=no --driveformat=virtio_scsi

(1/6) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.multi_disks.q35: STARTED
 (1/6) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.multi_disks.q35: ERROR: Login timeout expired    (output: "exceeded 360 s timeout, last failure: Client process terminated    (status: 1,    output: 'Ncat: TIMEOUT.\\n')") (416.64 s)
 (2/6) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.hotplug_unplug.q35: STARTED
 (2/6) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.hotplug_unplug.q35: ERROR: Login timeout expired    (output: "exceeded 360 s timeout, last failure: Client process terminated    (status: 1,    output: 'Ncat: TIMEOUT.\\n')") (414.08 s)
 (3/6) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.discard_check.q35: STARTED
 (3/6) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.discard_check.q35: PASS (7.66 s)
 (4/6) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.multi_disks.q35: STARTED
 (4/6) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.multi_disks.q35: ERROR: Shell command failed: 'D:\\coreutils\\DummyCMD.exe E:\\test.dat 102400000 1'    (status: -2146232576,    output: '\n') (113.20 s)
 (5/6) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.hotplug_unplug.q35: STARTED
 (5/6) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.hotplug_unplug.q35: ERROR: Shell command failed: 'D:\\coreutils\\DummyCMD.exe E:\\test.dat 102400000 1'    (status: -2146232576,    output: '\n') (111.62 s)
 (6/6) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.discard_check.q35: STARTED
 (6/6) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.discard_check.q35: PASS (7.62 s)

**Comment by qingwangrh** [comment]
> Hi @qingwangrh , hit some failed on win11 and win2022 guests:
> 
> python ConfigTest.py --testcase=vdpa_sim_blk_test --guestname=Win11,Win2022 --clone=no --driveformat=virtio_scsi
> 
> (1/6) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.multi_disks.q35: STARTED (1/6) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.multi_disks.q35: ERROR: Login timeout expired (output: "exceeded 360 s timeout, last failure: Client process terminated (status: 1, output: 'Ncat: TIMEOUT.\n')") (416.64 s) (2/6) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.hotplug_unplug.q35: STARTED (2/6) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.hotplug_unplug.q35: ERROR: Login timeout expired (output: "exceeded 360 s timeout, last failure: Client process terminated (status: 1, output: 'Ncat: TIMEOUT.\n')") (414.08 s) (3/6) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.discard_check.q35: STARTED (3/6) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.discard_check.q35: PASS (7.66 s) (4/6) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.multi_disks.q35: STARTED (4/6) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.multi_disks.q35: ERROR: Shell command failed: 'D:\coreutils\DummyCMD.exe E:\test.dat 102400000 1' (status: -2146232576, output: '\n') (113.20 s) (5/6) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.hotplug_unplug.q35: STARTED (5/6) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.hotplug_unplug.q35: ERROR: Shell command failed: 'D:\coreutils\DummyCMD.exe E:\test.dat 102400000 1' (status: -2146232576, output: '\n') (111.62 s) (6/6) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.discard_check.q35: STARTED (6/6) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.discard_check.q35: PASS (7.62 s)

I think failed case related your guest
1.Client process terminated (status: 1, output: 'Ncat: TIMEOUT.  
please make sure your guest is bootable, or your may try clone=yes

2.D:\coreutils\DummyCMD.exe E:\test.dat 
the guest miss .net 3.5 

https://bugzilla.redhat.com/show_bug.cgi?id=2218054 

I don't know why it exists your guest, maybe need to raise a new bug,  or you may change to other guests

**Comment by peixiu** [comment]
OK, I see the win11 has one jos succeed, so not doubt the guest status.
Any way, let me try again with clone=yes.

**Comment by peixiu** [comment]
(1/8) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: STARTED
 (1/8) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (914.44 s)
 (2/8) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: STARTED
 (2/8) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (684.85 s)
 (3/8) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.multi_disks.q35: STARTED
 (3/8) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.multi_disks.q35: PASS (126.64 s)
 (4/8) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.hotplug_unplug.q35: STARTED
 (4/8) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.hotplug_unplug.q35: PASS (172.19 s)
 (5/8) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.discard_check.q35: STARTED
 (5/8) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.discard_check.q35: PASS (7.68 s)
 (6/8) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.multi_disks.q35: STARTED
 (6/8) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.multi_disks.q35: PASS (97.76 s)
 (7/8) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.hotplug_unplug.q35: STARTED
 (7/8) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.hotplug_unplug.q35: PASS (168.07 s)
 (8/8) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.discard_check.q35: STARTED
 (8/8) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.vdpa_sim_blk_test.discard_check.q35: PASS (7.65 s)

LGTM~

File: provider/vdpa_sim_blk_utils.py [hunk 1]
```diff
@@ -75,16 +64,70 @@ def remove_vdpa_blk_dev(self, name):
...
+        cmd = "vdpa dev add mgmtdev vdpasim_blk name %s" % name
+        process.run(cmd, shell=True)
+        cmd = "vdpa dev list -jp %s" % name
+        process.run(cmd, shell=True)
+
+        time.sleep(2)
+        try:
+            dev = get_image_filename(name).replace("vdpa://", "")
+        except Exception as e:
+            raise VDPABlkSimulatorError("vdpa dev add %s failed:%s" % (name, e))
```
Comment (XueqiangWei) [inline provider/vdpa_sim_blk_utils.py:104]: Suggest to use str(e) instead of e. Thanks.

**Comment by XueqiangWei** [comment]
I think we need to add separate commit to the two files. like autotest/tp-qemu/pull/3497
e.g.
1. vdpa_sim_blk_utils.py
vdpa_sim_blk_utils: improve VDPABlkSimulatorTest

improve VDPABlkSimulatorTest to support virtio-vdpa and vhost-vdpa


2. vdpa_sim_blk_test.py
vdpa_sim_blk_test: point to new VirtioVdpaSimulatorTest

File: provider/vdpa_sim_blk_utils.py (hunk 1 above)
Comment (qingwangrh) [reply provider/vdpa_sim_blk_utils.py]: updated

**Review by XueqiangWei — APPROVED** [review]
